### ISSUE TITLE
The new 5.x.x Cocoa SDK is a little weird to use.

### ISSUE THREAD

**Issue by gabors** [getsentry/sentry-cocoa#516]
_Platform:_
- [x]  iOS
- [x]  tvOS
- [x]  MacOS
- [ ]  watchOS

_Swift:_
- [x]  Yes ->5.x
- [ ]  No

_sentry-cocoa installed with:_
- [x] SwiftPM

Version of sentry-cocoa:
**5.0.2**

---
I have following issue:

The new 5.x.x Cocoa SDK is a little weird to use.
You're supposed to init the SentrySDK using a string dictionary, **but valid keys and values are not documented anywhere that I can find**, apart from DSN. Prior version had properties on Client you can set.

How do I set environment, releaseName, max breadcrumbs, client level tags I was to set with every event? 

Is it expected to use the rather wordy syntax with building the Scope with every send???

```
SentrySDK.capture(message: "Test Message") { (scope) in
    scope.setEnvironment("staging")
    scope.setExtras(["ios": true])
    let u = Sentry.User(userId: "1")
    u.email = "tony@example.com"
    scope.setUser(u)
}
```

I didn't even got into capturing track trace and so on.

**Comment by CraigSiemens** [comment]
I agree with this. A dictionary based interface is much less user friendly than the previous property based interface.

I looks like there's an `Options` type that can the passed to start, but the only accessible init it has takes a dictionary as well. If `Options` had a public init with no params, then we could set the properties as needed before passing it to start, that'd be much nicer. `URLSession` and other types have a similar `initWithConfiguration` interface for setting them up and it works good.

**Comment by gabors** [comment]
Yeah this new SDK feels like a huge step backwards.
I ended up rolling back my apps to the previous one for the time being.

**Comment by philipphofmann** [comment]
Thanks, @gabors and @CraigSiemens for giving us feedback. I agree that the `Options` are not user friendly at the moment. If you want to know which parameters you can use checkout [`SentryOptions.m`](https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/SentryOptions.m#L52) at the moment. We are going to improve this to make it more user friendly. 
Do you have any other issues with the 5.x.x SDK at the moment or other things you wish were better?

**Comment by jkmassel** [comment]
I'll add on some (hopefully actionable) feedback:

Mirroring @CraigSiemens feedback – the dictionary approach for configuration is really hard to use (and may be a fatally flawed approach for some things – more on that in a moment). As a for-instance, it's really easy to make a typo and not understand why the configuration isn't working – that's an issue that should be caught at compile-time. One example of this is the fact that the properties in `SentryOptions.h` don't necessarily correspond to the dictionary keys – consider `releaseName` in `SentryOptions.h`. If you use that key, Sentry will silently replace it with its own default – [the code is expecting the key to be `release`](https://github.com/getsentry/sentry-cocoa/blob/854ddb4bca43cfd7c1119ec1bf0a9caa452bd74d/Sources/Sentry/SentryOptions.m#L87).

The other problem with using dictionaries is that storing a function in a dictionary doesn't work from Swift ([it looks like your team might've run into this already](https://github.com/getsentry/sentry-cocoa/blob/fc02d71795aa047382332dbc198aac40b59852fa/Tests/SentryTests/SentrySwiftTests.swift#L99)) – try changing [this line](https://github.com/getsentry/sentry-cocoa/blob/fc02d71795aa047382332dbc198aac40b59852fa/Tests/SentryTests/SentrySwiftTests.swift#L88) to the following to see the issue:
```swift
_ = SentrySDK(options: [
    "dsn": "https://username:password@app.getsentry.com/12345",
    "beforeSend": { debugPrint($0) }
])
```

If I may suggest it, your approach to event capture is a lot better and would be a great fit here. You could allow something like:

```swift
SentrySDK.start("__DSN__") { options in
  options.debug = true
  options.maxBreadcrumbs = 500
  options.enableAutoSessionTracking = true
  options.environment = "production"
  options.integrations = [.crash, .memoryWarning]
}
```

This would allow compiler-time error handling and make discoverability really simple. It would also allow backwards compatibility because you could add fields to the `Options` object with defaults and allow them to be overridden if needed as they're released. 

---

One other major area of difficulty I'm hitting trying to integrate the new SDK is the inability to detect that captured events have been uploaded successfully. [It seems like most of the work is done to allow this](https://github.com/getsentry/sentry-cocoa/blob/fc02d71795aa047382332dbc198aac40b59852fa/Sources/Sentry/SentryClient.m#L102), so I'm wondering if it's on the roadmap to allow passing in our own completion handler for this from `captureError`? Even if as a stopgap it was only available in `captureEvent:withScope:`, I'd be willing to do the work to assemble the event myself – this is unfortunately a blocker to us adopting the new SDK.

I hope this feedback is helpful and that this SDK continues to improve so that we can adopt it soon!

**Comment by mgray88** [comment]
Another issue is with the `User` object. There is no way to get the currently set object to be able to modify it. I ran into this on the Android SDK as well, and the solution there (https://github.com/getsentry/sentry-android/issues/385) was to use the `configureScope` with callback (which felt strange since it was accessing the same underlying object).

With this new SDK, I am not able to even access it that way. I need to be able to modify this object as the current user changes (goes from anonymous to logged in) and I shouldn't have to keep track of the `User` object outside of the SDK

**Comment by philipphofmann** [comment]
Thanks, @jkmassel and @mgray88 for your detailed feedback. We are going to try to use your feedback to improve the SDK.

**Comment by robertvoigt030** [comment]
We where using snapshotStacktrace in the pre 5.xx version to append stacktraces to our custom events. I couldn't find any compensational option within the new 5.xx version to enrich custom events with stacktraces. Is there any possibility to do so or any reason why we are not allowed to use this anymore?

**Comment by philipphofmann** [comment]
@robertvoigt030 yes we are currently working on it. One PR for it is almost done https://github.com/getsentry/sentry-cocoa/pull/545 and another one is in progress right now.

**Comment by ianthetechie** [comment]
Uhm, yes... I would use a slightly stronger word than "weird" myself. It makes no sense at all, frankly, and where on earth is the documentation? The migration guide covers the barebones, but is far from adequate. I was unable to find any docs besides the source on what keys you expect in the dict. This might be cool in JavaScript or some other ecosystem where stringly typed interfaces are basically your only option, but this is a huge step back for the Cocoa library.

**Comment by philipphofmann** [comment]
Thanks for your honest feedback @ianthetechie.

**Comment by SticklyMan** [comment]
I would agree that more complete documentation would help developers transition from the old to the new SDK with greater ease. Having directly accessible properties over arbitrary dictionary keys really just helps with the self-documenting code concept, plus it's always nice to have the compiler tell you when things are wrong or have changed.

FYI- We were able to switch our project over to the new SDK (not much changed in our implementation), but I did have to dig into the source code a few times to make sure we got things right. I would prefer that scope's setTag method be key first, followed by value, but that's just me.

I also want to draw attention to the fact that we are no longer able to detect if an event failed to send for whatever reason. We were using this to alert us if we passed bad values to our SentryEvent's extra data. Although our case is now likely covered by #509 (info would be missing rather than the entire report being missing), we may still like to have alerts just in case.

Thanks for all of the hard work on improving this library!
